Decouple testdata from bundle (#88)
operator-framework/enhancements: enhancements/testdata-separation.md

---
title: testdata-separation
authors:
  - "@theishshah"
reviewers:
  - TBD
approvers:
  - TBD
creation-date: 2021-07-12
last-updated: 2021-07-20
status: TBD
---

# testdata-separation

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA

## Summary

The current bundle includes both an operator as well as its testdata. This results in inflated binaries and scenarios where users may want to release an operator bundle sans testdata but are unable to do so. The solution to this issue is to provide users the ability to use and generate a slim bundle which is decoupled from the testing necessities. Due to the nature of the testdata, it needs the rest of the bundle so it has limited value alone.

## Motivation

> The motivation for this EP is to provide users with the option to deliver a slim bundle without testdata.

Reviewer: What do you mean with testdata? 
Are you speaking about the dir `[tests](https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v3/memcached-operator/bundle/tests)/scorecard/` ? Because we have not a testdata dir in the bundle. 

Is the goal remove the file : https://github.com/operator-framework/operator-sdk/blob/master/testdata/go/v3/memcached-operator/bundle/tests/scorecard/config.yaml? 

If yes, what is the size of this file? How much the authors will win with? 
Also, can we not only recommend dor authors delete the dir if they wish?

* Creates a smaller size bundle to ship
* Prevents including testing resources in shipped bundle
  * Stop potential elevated RBAC's in the test bundle from being included in production environments

### Goals

The goal is to create a solution which allows developers to continue to access and utilize the testdata while still having the option to ship an operator without said testdata in the bundle.

### Non-Goals

The goal of this EP is not to change the contents of the bundle nor the testdata, but just to reorganize them, either with multiple bundles or by allowing the tools to modify a bundle to achieve the desired state.

## Proposal

Please note that this proposed solution was elevated from the Alternatives selection after discussions, and the previous implementation has been moved down.

Hard Split

> * Split existing bundle in a testdata only bundle paired with a operator essentials only bundle

Reviewer 2: When a project is initialized, the `Makefile` is scaffolded with the assumption that there is a single bundle per project (see [here](https://github.com/operator-framework/operator-sdk/blob/2f6e5242a0a367feca82891d76283c537ee8fedb/internal/plugins/manifests/v2/init.go#L160-L186)). How will we overcome this assumption?

> * Split existing bundle in a testdata only bundle paired with a operator essentials only bundle

Reviewer 3: ```suggestion
* Split existing bundle in a testdata only bundle paired with an operator essentials only bundle
```
Nit typo :) but also what does it mean to be "operator essentials"?

  * Allows the bundle to save significant space/reduce its size

>   * Allows operator developers to securely ship with including potentially sensitive testdata

Reviewer 2: Will scorecard permit multiple testdata bundles to be referenced from an operator-only bundle? Consider the case where the operator authors decide some of their tests are sensitive, but other tests (e.g. the SDK provided tests) are not.

If multiple are allowed, can the main bundle include a scorecard configuration (like they do now) AND reference another bundle? 

If multiple are not allowed, what will we tell operator authors who want to separate tests out like this?

Reviewer 2: Another question. If the operator bundle is in a public registry repo and the testdata bundle is in a private registry repo, will callers of `operator-sdk scorecard` be expected to configure the credentials for the private testadata bundle repo out-of-band?

Should public operator bundles even reference private testdata bundles? If that is an intended use case, will SDK handle unauthorized errors gracefully? We can't assume that callers of `scorecard` will know in advance a) whether the main bundle includes its test config and b) if not, whether the referenced test config image is publicly available (or at least available to anyone who can pull the bundle itself).

Reviewer 3: @joelanford that is an EXCELLENT question because that's definitely one we have with `run bundle`.

    * Users can use elevated permissions in testing without concerns about shipping those configs
* Primary impact lies within scorecard internals, as well as scaffolding
  * Scorecard needs logic to understand whether or not there are two bundle or one
    * Recombine bundles at scorecard runtime
      * Scorecard runs continue to behave the same, all this forces is a pre-processing step

>     * Allows scorecard test developers to contiue to use `/bundle` context

Reviewer 3: ```suggestion
    * Allows scorecard test developers to continue to use `/bundle` context
```

* Allow discovery of test bundle using Annotation path

>   * Scorecard currently assumes testdata in the in current bundle, but can read the annotation to find another bundle

Reviewer 2: What happens if the testdata bundle referenced by the operator bundle does not exist?

If multiple references are allowed, what happens when some references exist and others don't?

>   * Scorecard currently assumes testdata in the in current bundle, but can read the annotation to find another bundle

Reviewer 3: ```suggestion
  * Scorecard currently assumes testdata is in the current bundle, but can read the annotation to find another bundle
```

>     * This bundle can be configured as the seperate test bundle

Reviewer 3: ```suggestion
    * This bundle can be configured as the separate test bundle
```

>   * Seamless tranisition for users, exisiting CLI commands will continue to work in scipts if annotations are correctly configured

Reviewer 3: ```suggestion
  * Seamless transition for users, existing CLI commands will continue to work in scripts if annotations are correctly configured
```

* Scorecard CLI extension
  * Add ability to process one OR MORE bundles in the command line args
  * If multiple args exist combine bundles and carry on
* Opt In design
  * By default bundles will continue to be scaffolded as is
  * Users can always re-scaffold the bundle to toggle between types
* Minimizes User Impact
  * Current operators will still work
  * New single bundle opertaors will not require any changes in the development or test process
    * Those who opt in will have minimal manual adjustments needed for the 2 bundles to work together
  * Scorecard test developers can expect the same scorecard API's and bundle mounting
 

## Design Details

### Test Plan

For testing, ideally all new paths created in the CLI should be verfied in integration testing. Creating a full bundle, slim bundle, and then adding and/or removing testdata scaffolds from both. 

### Graduation Criteria

**Note:** *Section not required until targeted at a release.*

Define graduation milestones.

Implementation and testing of this feature should be sufficient for graduation. The intended implementation will maintain current user behavior as default.

> 

Reviewer 2: Can we add a section that discusses the impact for disconnected clusters/users? Since bundles are required to be mirrored to enable OLM to deploy them in disconnected clusters, the testdata (for better or worse) comes along for the ride.

There are a few concerns to work through if we split up the testdata from the operator portion of the bundle:
1. How do we ensure that testdata-only bundles are included in the mirrored set of bundles? We need the testdata-only bundle to show up in the CSVs relatedImages field, right?
2. If operator authors choose not to ship their testdata-only bundles, are they expected to leave the operator bundle's test annotations out?
